File: test/ModularPipelines.UnitTests/Extensions/FolderExtensions.cs
using ModularPipelines.FileSystem;

namespace ModularPipelines.UnitTests.Extensions;

public static class FolderExtensions
{
    public static FolderPath? FindAncestorContainingProject(this FolderPath original)
    {
        var folder = original;

        while (folder != null)
        {
            if (folder.ListFiles().Any(x => x.Extension == ".csproj"))
            {
                return folder;
            }

            folder = folder.Parent;
        }

        return null;
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Extensions/FolderExtensions.fs
namespace ModularPipelines.UnitTests.FSharp.Extensions

open System.Linq
open ModularPipelines.FileSystem

module FolderExtensions =
    let findAncestorContainingProject (original: Folder) =
        let rec findProject (folder: Folder option) =
            match folder with
            | None -> None
            | Some f ->
                let hasProject = 
                    f.ListFiles()
                    |> Seq.exists (fun x -> x.Extension = ".csproj" || x.Extension = ".fsproj")
                if hasProject then
                    Some f
                else
                    findProject (Option.ofObj f.Parent)
        findProject (Some original)

type FolderExtensions =
    [<System.Runtime.CompilerServices.Extension>]
    static member FindAncestorContainingProject(original: Folder) =
        FolderExtensions.findAncestorContainingProject original
        |> Option.toObj
